lms/djangoapps/grades/tasks.py

```diff
@@ -19,25 +21,51 @@
 from .signals.signals import SUBSECTION_SCORE_CHANGED
 from .transformer import GradesTransformer
 
+log = getLogger(__name__)
+
 
 @task(default_retry_delay=30, routing_key=settings.RECALCULATE_GRADES_ROUTING_KEY)
-def recalculate_subsection_grade(user_id, course_id, usage_id, only_if_higher):
+def recalculate_subsection_grade(user_id, course_id, usage_id, only_if_higher, grade, max_grade):
     """
     Updates a saved subsection grade.
     This method expects the following parameters:
-       - user_id: serialized id of applicable User object
-       - course_id: Unicode string representing the course
-       - usage_id: Unicode string indicating the courseware instance
-       - only_if_higher: boolean indicating whether grades should
+        - user_id: serialized id of applicable User object
+        - course_id: Unicode string representing the course
+        - usage_id: Unicode string indicating the courseware instance
+        - only_if_higher: boolean indicating whether grades should
         be updated only if the new grade is higher than the previous
         value.
+        - grade: the points the learner earned on the problem that
+        triggered the update
+        - max_grade: the max points the leaner could have earned
+        on the problem
     """
     if not PersistentGradesEnabledFlag.feature_enabled(course_id):
         return
 
     course_key = CourseLocator.from_string(course_id)
     student = User.objects.get(id=user_id)
     scored_block_usage_key = UsageKey.from_string(usage_id).replace(course_key=course_key)
+    student_module = StudentModule.objects.get(
+        student_id=user_id,
+        module_state_key=scored_block_usage_key,
+        course_id=course_key,
+    )
+
+    # If the student module hadn't finished updating when the subsection update
+    # was initiated, the subsection update will attempt to use the
+    # learner's previous problem score, so we retry the subsection
+    # grade update here instead of proceeding.
+    grades_match = student_module.grade == grade or student_module.max_grade == max_grade
+
+    # We have to account for the situation where a student's state
+    # has been deleted- in this case, the saved score and max are None, but
+    # the score change will contain 0 for the grade and the max score
+    # for the problem for the max grade.
+    state_deleted = student_module.grade is None and student_module.max_grade is None and grade == 0
+
+    if not (state_deleted or grades_match):
+        _retry_recalculate_subsection_grade(user_id, course_id, usage_id, only_if_higher, grade, max_grade)
 
     collected_block_structure = get_course_in_cache(course_key)
```

Comment:
Please consider refactoring this method as it's getting long.  Perhaps a "simple change" like putting the rest of this method in a separate method.